pkg/kubelet/prober/runner.go

```diff
@@ -0,0 +1,255 @@
+	port    int
+	timeout time.Duration
+}
+
+// newGRPCProbeRunner returns grpcProbeRunner which implements probeRunner.
+func newGRPCProbeRunner() *grpcProbeRunner {
+	return &grpcProbeRunner{}
+}
+
+func (gp *grpcProbeRunner) setup(container v1.Container, status v1.PodStatus, probeType probeType) error {
+	probeSpec := getProbeSpecFromContainer(container, probeType)
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+
+	host := status.PodIP
+	service := ""
+	if probeSpec.GRPC.Service != nil {
+		service = *probeSpec.GRPC.Service
+	}
+
+	// persist
+	gp.host = host
+	gp.service = service
+	gp.port = int(probeSpec.GRPC.Port)
+	gp.timeout = timeout
+	return nil
+}
+
+func (gp *grpcProbeRunner) run(_ context.Context, prober prober) (probe.Result, string, error) {
+	return prober.grpc.Probe(gp.host, gp.service, gp.port, gp.timeout)
+}
+
+// execProbeRunner holds state and logic for running Exec probes.
+type execProbeRunner struct {
+	containerID kubecontainer.ContainerID
+	command     []string
+	timeout     time.Duration
+}
+
+// newExecProbeRunner returns execProbeRunner which implements probeRunner.
+func newExecProbeRunner() *execProbeRunner {
+	return &execProbeRunner{}
+}
+
+func (ep *execProbeRunner) setup(container v1.Container, status v1.PodStatus, probeType probeType) error {
+	probeSpec := getProbeSpecFromContainer(container, probeType)
+	command := kubecontainer.ExpandContainerCommandOnlyStatic(probeSpec.Exec.Command, container.Env)
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+
+	c, _ := podutil.GetContainerStatus(status.ContainerStatuses, container.Name)
+	containerID := kubecontainer.ParseContainerID(c.ContainerID)
+
+	// persist
+	ep.command = command
+	ep.timeout = timeout
+	ep.containerID = containerID
+	return nil
+}
+
+func (ep *execProbeRunner) run(ctx context.Context, prober prober) (probe.Result, string, error) {
+	return prober.exec.Probe(&execInContainer{run: func() ([]byte, error) {
+		return prober.runner.RunInContainer(ctx, ep.containerID, ep.command, ep.timeout)
+	}})
+}
+
+// emptyProbeRunner holds state and logic for running Exec probes.
+type emptyProbeRunner struct {
+	isProbeSpecNil   bool
+	isHandlerMissing bool
+}
+
+// newEmptyProbeRunner returns emptyProbeRunner which implements probeRunner.
+func newEmptyProbeRunner(isProbeSpecNil, isHandlerMissing bool) *emptyProbeRunner {
+	return &emptyProbeRunner{
+		isProbeSpecNil:   isProbeSpecNil,
+		isHandlerMissing: isHandlerMissing,
+	}
+}
+
+func (ep *emptyProbeRunner) setup(_ v1.Container, _ v1.PodStatus, _ probeType) error {
+	return nil
+}
+
+func (ep *emptyProbeRunner) run(_ context.Context, _ prober) (probe.Result, string, error) {
+	// waiting for development around https://github.com/kubernetes/kubernetes/issues/116026
+	// to finalize where warning should be emitted
+	if ep.isProbeSpecNil {
+		return probe.Success, "", nil
+	}
+
+	if ep.isHandlerMissing {
+		return probe.Unknown, "", fmt.Errorf("missing probe handler")
```

Comment:
oh, ok. Than this will be covered by that other task that will expose this error as an event. Makes sense